File: apps/api/app/services/environment_readiness.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any


@dataclass
class ChecklistTemplateItem:
    environment: str
    item_key: str
    label: str
    owner: str
    category: str
    required: bool = True
    note: str | None = None


def checklist_template() -> list[ChecklistTemplateItem]:
    return [
        ChecklistTemplateItem("PREVIEW", "preview.runtime", "Preview runtime orchestration", "platform", "runtime"),
        ChecklistTemplateItem("PREVIEW", "preview.domain", "Temporary preview URL provisioning", "platform", "deployment"),
        ChecklistTemplateItem("PREVIEW", "preview.recovery", "Preview recovery and restart", "platform", "recovery"),
        ChecklistTemplateItem("PREVIEW", "preview.provider", "Connect deployment provider (Vercel/Render)", "user", "deployment"),
        ChecklistTemplateItem("PREVIEW", "preview.repo", "Connect repository for deployment bootstrap", "user", "deployment"),
        ChecklistTemplateItem("STAGING", "staging.orchestration", "Staging deployment orchestration", "platform", "deployment"),
        ChecklistTemplateItem("STAGING", "staging.rollback", "Rollback workflow available", "platform", "recovery"),
        ChecklistTemplateItem("STAGING", "staging.secrets", "Provide staging environment secrets", "user", "secrets"),
        ChecklistTemplateItem("STAGING", "staging.auth", "Configure staging auth callbacks/domains", "user", "auth"),
        ChecklistTemplateItem("PRODUCTION", "production.governance", "Promotion and deployment governance", "platform", "governance"),
        ChecklistTemplateItem("PRODUCTION", "production.recovery", "Rollback and recovery orchestration", "platform", "recovery"),
        ChecklistTemplateItem("PRODUCTION", "production.domain", "Configure custom domain and DNS", "user", "deployment"),
        ChecklistTemplateItem("PRODUCTION", "production.secrets", "Set production secrets", "user", "secrets"),
        ChecklistTemplateItem("PRODUCTION", "production.monitoring", "Enable production monitoring", "user", "operations"),
        ChecklistTemplateItem("PRODUCTION", "production.backup", "Define backup and restore policy", "user", "operations"),
    ]


def infer_item_status(
    template: ChecklistTemplateItem,
    *,
    has_repo: bool,
    has_connector: bool,
    preview_ready: bool,
    foundation_missing: set[str],
) -> tuple[str, str | None]:
    key = template.item_key
    if template.owner == "platform":
        if key == "preview.runtime" and not preview_ready:
            return "pending", "No healthy preview run verified yet."
        return "done", None

    if key.endswith("provider"):
        return ("done", None) if has_connector else ("pending", "Connect Vercel or Render provider.")
    if key.endswith("repo"):
        return ("done", None) if has_repo else ("pending", "Repository must be connected.")
    if "auth" in key and any("auth" in item for item in foundation_missing):
        return "pending", "Auth prerequisites are incomplete."
    if "secrets" in key and any(item in foundation_missing for item in {"architecture", "repo", "preview"}):
        return "pending", "Foundation prerequisites should be completed before secrets handoff."
    if key in {"production.monitoring", "production.backup"}:
        return "pending", template.note or "User action required for production operations."
    return "done", None


def complete_timestamp(status: str) -> datetime | None:
    if status != "done":
        return None
    return datetime.now(timezone.utc)


def normalize_missing_prerequisites(values: list[str] | None) -> set[str]:
    if not values:
        return set()
    return {str(value or "").strip().lower() for value in values if str(value or "").strip()}


def summarize_rows(rows: list[Any]) -> dict[str, Any]:
    grouped: dict[str, dict[str, Any]] = {}
    for env in ("PREVIEW", "STAGING", "PRODUCTION"):
        grouped[env] = {
            "environment": env,
            "total": 0,
            "completed": 0,
            "platform_total": 0,
            "platform_completed": 0,
            "user_pending": 0,
        }
    for row in rows:
        env = str(getattr(row, "environment", "")).upper()
        if env not in grouped:
            grouped[env] = {
                "environment": env,
                "total": 0,
                "completed": 0,
                "platform_total": 0,
                "platform_completed": 0,
                "user_pending": 0,
            }
        bucket = grouped[env]
        owner = str(getattr(row, "owner", "user")).lower()
        status = str(getattr(row, "status", "pending")).lower()
        bucket["total"] += 1
        if status == "done":
            bucket["completed"] += 1
        if owner == "platform":
            bucket["platform_total"] += 1
            if status == "done":
                bucket["platform_completed"] += 1
        if owner == "user" and status != "done":
            bucket["user_pending"] += 1

    envs = []
    total = 0
    completed = 0
    for env in ("PREVIEW", "STAGING", "PRODUCTION"):
        bucket = grouped[env]
        bucket_total = bucket["total"]
        bucket["score_pct"] = int(round((bucket["completed"] / bucket_total) * 100)) if bucket_total else 0
        total += bucket_total
        completed += bucket["completed"]
        envs.append(bucket)
    score_pct = int(round((completed / total) * 100)) if total else 0
    return {
        "score_pct": score_pct,
        "environments": envs,
        "total": total,
        "completed": completed,
    }

File: apps/web/src/composables/environmentReadiness.ts
export type EnvironmentType = "PREVIEW" | "STAGING" | "PRODUCTION";
export type ReadinessOwner = "platform" | "user";
export type ReadinessStatus = "done" | "pending";

export type EnvironmentChecklistItem = {
  id: string;
  label: string;
  owner: ReadinessOwner;
  status: ReadinessStatus;
  note?: string;
};

export type EnvironmentChecklistSummary = {
  environment: EnvironmentType;
  scorePct: number;
  total: number;
  completed: number;
  platformCompleted: number;
  platformTotal: number;
  userPending: number;
  items: EnvironmentChecklistItem[];
};

export type EnvironmentReadinessInput = {
  hasRepo: boolean;
  hasDeploymentConnector: boolean;
  deploymentProviders?: string[];
  foundationMissing?: string[];
  previewReady?: boolean;
  deploymentPreflightOk?: boolean | null;
};

export function buildEnvironmentReadiness(input: EnvironmentReadinessInput) {
  const providers = (input.deploymentProviders || []).map((value) => String(value || "").toLowerCase()).filter(Boolean);
  const hasVercel = providers.includes("vercel");
  const hasRender = providers.includes("render");
  const missing = new Set((input.foundationMissing || []).map((value) => String(value || "").toLowerCase()));
  const hasAuthMissing = [...missing].some((row) => row.includes("auth"));
  const hasPreviewMissing = [...missing].some((row) => row.includes("preview"));
  const hasRepoMissing = [...missing].some((row) => row.includes("repo"));
  const hasArchMissing = [...missing].some((row) => row.includes("arch"));
  const previewVerified = Boolean(input.previewReady || input.deploymentPreflightOk);

  const preview = summarize("PREVIEW", [
    platform("preview_runtime", "Preview runtime orchestration", true),
    platform("preview_domain", "Temporary preview URL provisioning", true),
    platform("preview_recovery", "Preview restart and recovery handling", true),
    platform("preview_health", "Preview health verification", previewVerified, previewVerified ? "" : "No active healthy preview verified yet."),
    user("preview_connector", "Connect deployment provider (Vercel/Render)", input.hasDeploymentConnector),
    user("preview_repo", "Connect repository for deployment bootstrap", input.hasRepo && !hasRepoMissing),
  ]);

  const staging = summarize("STAGING", [
    platform("staging_deploy_flow", "Staging deployment orchestration", true),
    platform("staging_retry", "Retry/rollback workflow availability", true),
    user("staging_secrets", "Provide staging environment secrets", !hasArchMissing, hasArchMissing ? "Architecture/foundation prerequisites still missing." : ""),
    user("staging_connectors", "Connect at least one deployment provider", input.hasDeploymentConnector),
    user("staging_auth", "Configure staging auth callbacks/domains", !hasAuthMissing, hasAuthMissing ? "Auth-related prerequisite is still missing." : ""),
    user("staging_checks", "Run staging integration checks", !hasPreviewMissing, hasPreviewMissing ? "Preview prerequisites suggest staging checks are incomplete." : ""),
  ]);

  const production = summarize("PRODUCTION", [
    platform("prod_governance", "Deployment governance and promotion controls", true),
    platform("prod_recovery", "Rollback + recovery orchestration", true),
    user("prod_domain", "Configure production domain and DNS", hasVercel || hasRender, "Custom domain and DNS are user-owned."),
    user("prod_secrets", "Set production secrets (DB/auth/payments)", !hasArchMissing && input.hasRepo, "Credentials and rotation remain user-owned."),
    user("prod_monitoring", "Enable production monitoring and alerting", false, "Monitoring ownership remains with workspace operators."),
    user("prod_backup", "Define backup and restore policy", false, "Backup/compliance policy is required before production trust is high."),
  ]);

  const all = [preview, staging, production];
  const total = all.reduce((acc, env) => acc + env.total, 0);
  const completed = all.reduce((acc, env) => acc + env.completed, 0);
  const scorePct = total ? Math.round((completed / total) * 100) : 0;

  return {
    scorePct,
    environments: all,
    nextUserActions: all.flatMap((env) => env.items.filter((item) => item.owner === "user" && item.status === "pending")).slice(0, 6),
  };
}

function summarize(environment: EnvironmentType, items: EnvironmentChecklistItem[]): EnvironmentChecklistSummary {
  const total = items.length;
  const completed = items.filter((item) => item.status === "done").length;
  const platformItems = items.filter((item) => item.owner === "platform");
  const platformCompleted = platformItems.filter((item) => item.status === "done").length;
  const userPending = items.filter((item) => item.owner === "user" && item.status === "pending").length;
  return {
    environment,
    total,
    completed,
    platformCompleted,
    platformTotal: platformItems.length,
    userPending,
    scorePct: total ? Math.round((completed / total) * 100) : 0,
    items,
  };
}

function platform(id: string, label: string, done: boolean, note = ""): EnvironmentChecklistItem {
  return { id, label, owner: "platform", status: done ? "done" : "pending", note: note || undefined };
}

function user(id: string, label: string, done: boolean, note = ""): EnvironmentChecklistItem {
  return { id, label, owner: "user", status: done ? "done" : "pending", note: note || undefined };
}
